File: MATLAB/TDOA Localization code/TDOA Localization/Iterative Algorithm (IA)/PSO.m
clear
clc

%% 1. Call Sensors_Data - Contained: 4 sensor coordinates, Speed of EM wave, PD TDOA & PD coordinates (for benchmarking).
[receiver1, receiver2, receiver3, receiver4, speed_of_EM, data_predict] = func_Sensors_Data();

% Loop through all the TDOA in data_predict
num_of_loop = size(data_predict, 1);

Computation_Time = zeros(num_of_loop, 1);
FOR_EXCEL = cell(num_of_loop, 5);


% trial

for data_set = 1:num_of_loop
    StartTime = datetime('now');

    %% 3. Perform Sorting_data_predict: sorting the respective PD TDOA and coordinate in the data_predict list
    [true_source_location, TDOA_N] = func_Sorting_data_predict(data_predict, data_set);

    % PSO START HERE:
    
    % Parameters for PSO and convergence criterion
    max_iterations_pso = 50000;
    convergence_threshold = 1e-12; % Threshold for change in global best value
    convergence_iter_threshold = 1000; % Number of consecutive iterations with no improvement to trigger convergence

    convergence_counter = 0; % Initialize convergence counter
    
    global_best_value = inf;    prev_global_best_value = 0;    objective_values = zeros; % To clear current data_set and ready for next data_set

    % Initialization
    num_particles = 1000;
    inertia_weight = 0.1;
    c1 = 3.0; % Cognitive coefficient
    c2 = 1.0; % Social coefficient

    % Initialize particles' positions and velocities *within boundary
    particles_position = zeros(num_particles, 3);    particles_velocity = zeros(num_particles, 3);    
    for i = 1:num_particles
        % particles_position(i, 1) = rand() * 26 - 15; % num_particles(:, 1) range from -15 to 11
        particles_position(i, 1) = rand() * 11 + 0.5; % num_particles(:, 1) range from 1 to 11
        particles_position(i, 2) = rand() *11 + 0.5; % num_particles(:, 2) range between 1 to 11
        particles_position(i, 3) = rand() * 6 + 0.1; % num_particles(:, 3) range between 0.3 to 2
        
        particles_velocity(i, :) = rand(1, 3); % Initialize velocities randomly
    end

    objective_values_particles = zeros(num_particles, 1);

    % Initialize personal best positions and the corresponding objective values
    personal_best_positions = particles_position;    personal_best_values = ones(num_particles, 1);
    
    % Initialize the global best position and the corresponding objective value
    global_best_position = particles_position(1, :);  

    % Initialize the plot -> For Plotting Live Objective Value
    % figure;
    % h = plot(NaN, NaN, 'LineWidth', 2);
    % xlabel('Iterations');
    % ylabel('Global Best Value');
    % title('Global Best Value vs Iteration');
    % grid on;

    %% 9. Display_PSO_Moving_Particles
    % func_Display_PSO_Moving_Particles(global_best_position, particles_position, num_particles);

    for iter_pso = 1:max_iterations_pso
    
        % Evaluate the FITNESS of each particle
        for i = 1:num_particles
            x0 = particles_position(i, 1);
            y0 = particles_position(i, 2);
            z0 = particles_position(i, 3);

            %% 4. Objective_Function: Formulate of c = d/t
            [r1, r2, r3] = func_Objective_Function(x0, y0, z0, receiver1, receiver2, receiver3, receiver4, speed_of_EM, TDOA_N);
                                  
            % Calculate the objective value (sum of squared residuals)
            objective_values_particles(i) = sqrt((r1^2 + r2^2 + r3^2)/3); % RMSE

            % Update personal best if the current position is better
            if objective_values_particles(i) < personal_best_values(i)
                personal_best_positions(i, :) = particles_position(i, :);
                personal_best_values(i) = objective_values_particles(i);
            end
            
            % Update global best if the current position is better
            if objective_values_particles(i) < global_best_value
                global_best_position = particles_position(i, :);
                global_best_value = objective_values_particles(i);
            end
             
            % Update velocity
            particles_velocity(i, :) = inertia_weight * particles_velocity(i, :) + ...
                c1 * rand() * (personal_best_positions(i, :) - particles_position(i, :)) + ...
                c2 * rand() * (global_best_position - particles_position(i, :));
            
            % Update position
            particles_position(i, :) = particles_position(i, :) + particles_velocity(i, :);

            % Bound positions within search space [1, 1, 0] (lower bounds) + [11, 11, 1.8] (upper bounds)
            particles_position(i, :) = max(min(particles_position(i, :), [11 11 6]), [0.5 0.5 0.1]); %<=======================
            % particles_position(i, :) = max(min(particles_position(i, :), [11 11 2]), [-15 1 0.3]);

            objective_values(iter_pso, 1) = global_best_value;
        end        
        %% 5. Convergence_check: stop running when convergence met
        [convergence_counter] = func_Convergence_Check(iter_pso, global_best_value, prev_global_best_value, convergence_threshold, convergence_counter);
               
        % Check convergence criterion
        if convergence_counter >= convergence_iter_threshold
            fprintf('Convergence did not improve after %d iterations (iter_pso = %d).\n', convergence_iter_threshold, iter_pso);
            break; % Exit the PSO loop if convergence criterion is met
        end               
        
        % Store current global best value for next iteration
        prev_global_best_value = global_best_value;

        % %% For Plotting Live Objective Value
        % % Update the plot with the new data
        % set(h, 'XData', 1:iter_pso, 'YData', objective_values(1:iter_pso));
        % 
        % % Pause to allow the plot to update
        % pause(0.1); % Adjust pause duration as needed

        %% 10. Update_PSO_Moving_Particles
        % func_Update_PSO_Moving_Particles(global_best_position, particles_position, num_particles)

    end    

    % Final estimated source location from PSO
    estimated_location = global_best_position;   
    
    %% 6. Compute_Error
    [Euclidean_Distance, percentage_error] = func_Compute_Error(true_source_location, estimated_location, receiver1, receiver2, receiver3, receiver4);
    
    %% 7. Display_Convergence_Plot
    % func_Display_Convergence_Plot(iter_pso, objective_values);
    
    %% 8. Display_results
    func_Display_Results(estimated_location, true_source_location, Euclidean_Distance, percentage_error);

    EndTime = datetime('now');

    durationBetween = EndTime - StartTime;
    
    % Convert the duration to seconds before assigning
    % Computation_Time(data_set, 1) = seconds(durationBetween);

    % Round off for coordinate only before copied to excel
    % estimated_location = arrayfun(@(x) sprintf('%.2f', x), estimated_location, 'UniformOutput', false);
    true_source_location = arrayfun(@(x) sprintf('%.2f', x), true_source_location, 'UniformOutput', false);

    % Format Numbers as Strings Before Export
    % With round off
    % Round to 4 significant figures
    FOR_EXCEL{data_set, 1} = round(percentage_error, 2);
    FOR_EXCEL{data_set, 2} = round(Euclidean_Distance, 3);
    FOR_EXCEL{data_set, 3} = round(seconds(durationBetween), 2);
    FOR_EXCEL{data_set, 4} = round(estimated_location(1), 2);
    FOR_EXCEL{data_set, 5} = round(estimated_location(2), 2);
    FOR_EXCEL{data_set, 6} = round(estimated_location(3), 2);
    FOR_EXCEL{data_set, 7} = strjoin(true_source_location, ', ');
end

% Save the data
save('FOR_EXCEL_PSO.mat', 'FOR_EXCEL');

writecell(FOR_EXCEL, 'converted_data.xlsx');

File: (Deprecated) Python/(Deprecated) PSO/PSO.py
import numpy as np
from datetime import datetime
from typing import Tuple, List
import time

def sensors_data() -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, float, np.ndarray]:
    """Initialize sensor positions and TDOA measurement data."""
    # Speed of EM Wave
    speed_of_EM = 299792458

    # Sensor placement
    receiver1 = np.array([1, 3, 1.4])
    receiver2 = np.array([1, 5, 1.8])
    receiver3 = np.array([1, 7, 1.4])
    receiver4 = np.array([1, 9, 1.8])

    # Data predict format: X, Y, Z, T21, T31, T41
    # Time unit is in ns
    data_predict = np.array([
        # location 2
        # [1.999993 + 1, -0.0122108 + 1, 1.2, 5.3838757, 11.3425882, 17.8626851]
        [2.999993, 0.9877892, 1.2, 5.3838757, 11.3425882, 17.8626851],
        # Add more data points as needed
    ])

    return receiver1, receiver2, receiver3, receiver4, speed_of_EM, data_predict

def sorting_data_predict(data_predict: np.ndarray, data_set: int) -> Tuple[np.ndarray, np.ndarray]:
    """Extract true source location and TDOA measurements for a specific dataset."""
    true_source_location = data_predict[data_set-1, :3]
    TDOA_N = (data_predict[data_set-1, 3:])*1e-9  # Convert to seconds
    return true_source_location, TDOA_N

def objective_function(position: np.ndarray, receivers: List[np.ndarray], speed_of_EM: float, TDOA_N: np.ndarray) -> Tuple[float, float, float, float, float, float, float]:
    """Calculate objective function value for PSO."""
    x0, y0, z0 = position
    
    # Calculate distances to each receiver
    d1 = np.sqrt(np.sum((position - receivers[0])**2))
    d2 = np.sqrt(np.sum((position - receivers[1])**2))
    d3 = np.sqrt(np.sum((position - receivers[2])**2))
    d4 = np.sqrt(np.sum((position - receivers[3])**2))

    # Calculate residuals
    r1 = d2 - d1 - (speed_of_EM * TDOA_N[0])
    r2 = d3 - d1 - (speed_of_EM * TDOA_N[1])
    r3 = d4 - d1 - (speed_of_EM * TDOA_N[2])
    
    return r1, r2, r3, d1, d2, d3, d4

def compute_error(true_source_location: np.ndarray, estimated_location: np.ndarray, receivers: List[np.ndarray]) -> Tuple[float, float]:
    """Compute error metrics between true and estimated locations."""
    euclidean_distance = np.sqrt(np.sum((true_source_location - estimated_location)**2))
    # Calculate percentage error (assuming maximum possible error is based on domain size)
    sum_total = np.linalg.norm((np.abs(estimated_location - receivers[0]) + np.abs(estimated_location - receivers[1]) + np.abs(estimated_location - receivers[2]) + np.abs(estimated_location - receivers[3]))/4)
    percentage_error = (euclidean_distance / sum_total) * 100
    return euclidean_distance, percentage_error

def pso_algorithm():
    """Main PSO algorithm implementation."""
    # Get sensor data
    receiver1, receiver2, receiver3, receiver4, speed_of_EM, data_predict = sensors_data()
    receivers = [receiver1, receiver2, receiver3, receiver4]
    
    num_of_loop = len(data_predict)
    results = []

    for data_set in range(1, num_of_loop + 1):
        start_time = datetime.now()

        # Get true source location and TDOA measurements
        true_source_location, TDOA_N = sorting_data_predict(data_predict, data_set)

        # PSO parameters
        max_iterations_pso = 50000
        convergence_threshold = 1e-12
        convergence_iter_threshold = 1000
        num_particles = 1000
        inertia_weight = 0.1
        c1 = 3.0  # Cognitive coefficient
        c2 = 1.0  # Social coefficient

        # Initialize particles within boundary
        particles_position = np.random.uniform(
            low=[0.5, 0.5, 0.1],
            high=[11, 11, 2],
            size=(num_particles, 3)
        )
        particles_velocity = np.random.rand(num_particles, 3)
        
        # Initialize best positions and values
        personal_best_positions = particles_position.copy()
        personal_best_values = np.ones(num_particles) * np.inf
        global_best_value = np.inf
        global_best_position = particles_position[0].copy()
        
        convergence_counter = 0
        prev_global_best_value = 0

        # Main PSO loop
        for iter_pso in range(max_iterations_pso):
            for i in range(num_particles):
                # Calculate objective function
                r1, r2, r3, *_ = objective_function(
                    particles_position[i],
                    receivers,
                    speed_of_EM,
                    TDOA_N
                )
                
                # Calculate RMSE
                objective_value = np.sqrt((r1**2 + r2**2 + r3**2)/3)
                
                # Update personal best
                if objective_value < personal_best_values[i]:
                    personal_best_values[i] = objective_value
                    personal_best_positions[i] = particles_position[i].copy()
                
                # Update global best
                if objective_value < global_best_value:
                    global_best_value = objective_value
                    global_best_position = particles_position[i].copy()
            
            # Update velocities and positions
            r1, r2 = np.random.rand(num_particles, 1), np.random.rand(num_particles, 1)
            particles_velocity = (inertia_weight * particles_velocity +
                               c1 * r1 * (personal_best_positions - particles_position) +
                               c2 * r2 * (global_best_position - particles_position))
            
            particles_position += particles_velocity
            
            # Apply boundary conditions
            particles_position = np.clip(particles_position, [0.5, 0.5, 0.1], [11, 11, 2])
            
            # Check convergence
            if abs(global_best_value - prev_global_best_value) < convergence_threshold:
                convergence_counter += 1
            else:
                convergence_counter = 0
            
            if convergence_counter >= convergence_iter_threshold:
                print(f'Convergence reached after {iter_pso} iterations.')
                break
            
            prev_global_best_value = global_best_value
        
        # Compute final errors
        euclidean_distance, percentage_error = compute_error(true_source_location, global_best_position, receivers)
        
        end_time = datetime.now()
        duration = (end_time - start_time).total_seconds()
        
        # Store results
        results.append({
            'data_set': data_set,
            'percentage_error': round(percentage_error, 2),
            'euclidean_distance': round(euclidean_distance, 3),
            'computation_time': round(duration, 2),
            'estimated_location': [round(x, 2) for x in global_best_position],
            'true_location': [round(x, 2) for x in true_source_location]
        })
        
        # Display results
        print(f"\nResults for dataset {data_set}:")
        print(f"Estimated location: {results[-1]['estimated_location']}")
        print(f"True location: {results[-1]['true_location']}")
        print(f"Euclidean distance error: {results[-1]['euclidean_distance']}")
        print(f"Percentage error: {results[-1]['percentage_error']}%")
        print(f"Computation time: {results[-1]['computation_time']} seconds")

    return results

if __name__ == "__main__":
    results = pso_algorithm()